### ISSUE TITLE
protocol might not work for functions require multiple cross-chain messages

### ISSUE THREAD

**Issue by c4-submissions** [code-423n4/2023-09-maia-findings#736]
# Lines of code

https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchBridgeAgent.sol#L276-L303
https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/RootBridgeAgent.sol#L577-L617

# Vulnerability details

## Impact
There are some functions require multiple cross-chain messages between Branch chain and Root chain, but while transfer ETH to LayerZero for sending messages cross-chain, LayerZero requires each messages should contains fees, otherwise, the call will revert with [LayerZero: not enough native for fees](https://layerzero.gitbook.io/docs/evm-guides/error-messages/failure-revert-messages#layerzero-not-enough-native-for-fees), but when handling multiple cross-chain messages, the system has some issues, which might cause:
1. If the protocol have ETH within the pool, then user's tx will steal all the ETH
1. If the protocol doesn't have any ETH, then user's tx will revert

## Proof of Concept
Take [BranchBridgeAgent.callOutSignedAndBridge](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchBridgeAgent.sol#L276-L303) as an example, while a user Alice calls the function with properly parameters on Branch chain, and she sends enough gas for `BranchBridgeAgent.callOutSignedAndBridge`, the tx is executed successfully, and the messages will cross chain to Root branch.
In Root chain, [RootBridgeAgent.sol#L577-L617](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/RootBridgeAgent.sol#L577-L617) will handle Alice's message,
```solidity
 577         } else if (_payload[0] & 0x7F == 0x05) {
 578             // Parse deposit nonce
 579             nonce = uint32(bytes4(_payload[PARAMS_START_SIGNED:PARAMS_TKN_START_SIGNED]));
 580 
 581             //Check if tx has already been executed
 582             if (executionState[_srcChainId][nonce] != STATUS_READY) {
 583                 revert AlreadyExecutedTransaction();
 584             }
 585 
 586             // Get User Virtual Account
 587             VirtualAccount userAccount = IPort(localPortAddress).fetchVirtualAccount(
 588                 address(uint160(bytes20(_payload[PARAMS_START:PARAMS_START_SIGNED])))
 589             );
 590 
 591             // Toggle Router Virtual Account use for tx execution
 592             IPort(localPortAddress).toggleVirtualAccountApproved(userAccount, localRouterAddress);
 593 
 594             // Avoid stack too deep
 595             uint16 srcChainId = _srcChainId;
 596 
 597             // Try to execute remote request
 598             // Flag 5 - RootBridgeAgentExecutor(bridgeAgentExecutorAddress).executeSignedWithDeposit(address(userAccount), localRouterAddress, data, _srcChainId)
 599             _execute(
 600                 _payload[0] == 0x85,
 601                 nonce,
 602                 address(uint160(bytes20(_payload[PARAMS_START:PARAMS_START_SIGNED]))),
 603                 abi.encodeWithSelector(
 604                     RootBridgeAgentExecutor.executeSignedWithDeposit.selector,
 605                     address(userAccount), <------- Here is Alice's VirtualAccount address
 606                     localRouterAddress,
 607                     _payload,
 608                     srcChainId
 609                 ),
 610                 srcChainId
 611             );
 612 
 613             // Toggle Router Virtual Account use for tx execution
 614             IPort(localPortAddress).toggleVirtualAccountApproved(userAccount, localRouterAddress);
 615 
 616             // DEPOSIT FLAG: 6 (Call with multiple asset Deposit + msg.sender)

```
In [RootBridgeAgent.sol#L605](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/RootBridgeAgent.sol#L605), **userAccount is Alice's VirtualAccount address**.

And then [RootBridgeAgent._execute](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/RootBridgeAgent.sol#L768-L794) is called
```solidity
 768     function _execute(
 769         bool _hasFallbackToggled,
 770         uint32 _depositNonce,
 771         address _refundee,
 772         bytes memory _calldata,
 773         uint16 _srcChainId
 774     ) private {
 775         //Update tx state as executed
 776         executionState[_srcChainId][_depositNonce] = STATUS_DONE;
 777 
 778         //Try to execute the remote request
 779         (bool success,) = bridgeAgentExecutorAddress.call{value: address(this).balance}(_calldata);     <--------- Here Alice might not send ETH to the RootBridgeAgent contract
 780 
 781         //Update tx state if execution failed
 782         if (!success) {
 783             //Read the fallback flag.
 784             if (_hasFallbackToggled) {
 785                 // Update tx state as retrieve only
 786                 executionState[_srcChainId][_depositNonce] = STATUS_RETRIEVE;
 787                 // Perform the fallback call
 788                 _performFallbackCall(payable(_refundee), _depositNonce, _srcChainId);
 789             } else {
 790                 // No fallback is requested revert allowing for retry.
 791                 revert ExecutionFailure();
 792             }
 793         }
 794     }
```
As the comment shows in [RootBridgeAgent.sol#L779](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/RootBridgeAgent.sol#L779), Alice might not send ETH to `RootBridgeAgent`, so the balance of `RootBridgeAgent` is
1. **0** 
1. **non-zero**: in such case, it means the protocol uses other's ETH

Continuing the code flow, after calling `RootBridgeAgent._execute`, the code arrives at [RootBridgeAgentExecutor.executeSignedWithDeposit](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/RootBridgeAgentExecutor.sol#L167-L191), supposed that `_payload.length > PARAMS_SETTLEMENT_OFFSET`, so the code in [RootBridgeAgentExecutor.sol#L185-L190](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/RootBridgeAgentExecutor.sol#L185-L190) will get executed, in [L187-L189](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/RootBridgeAgentExecutor.sol#L187-L189), the ETH will sends to `_router`
```solidity
167     function executeSignedWithDeposit(address _account, address _router, bytes calldata _payload, uint16 _srcChainId)
168         external
169         payable
170         onlyOwner
171     {
172         //Read Deposit Params
173         DepositParams memory dParams = DepositParams({
174             depositNonce: uint32(bytes4(_payload[PARAMS_START_SIGNED:PARAMS_TKN_START_SIGNED])),
175             hToken: address(uint160(bytes20(_payload[PARAMS_TKN_START_SIGNED:45]))),
176             token: address(uint160(bytes20(_payload[45:65]))),
177             amount: uint256(bytes32(_payload[65:97])),
178             deposit: uint256(bytes32(_payload[97:PARAMS_SETTLEMENT_OFFSET]))
179         });
180 
181         //Bridge In Asset
182         _bridgeIn(_account, dParams, _srcChainId);
183 
184         // Check if there is additional calldata in the payload
185         if (_payload.length > PARAMS_SETTLEMENT_OFFSET) {
186             //Execute remote request
187             IRouter(_router).executeSignedDepositSingle{value: msg.value}( <------- Here sending ETH to _router
188                 _payload[PARAMS_SETTLEMENT_OFFSET:], dParams, _account, _srcChainId
189             );
190         }
191     }

```
Two things to be noticed:
1. the `_account` parameter in `executeSignedWithDeposit` will be Alice's VirtualAccount address
1. in `RootBridgeAgentExecutor.sol`, the contract will send all  its ETH plus ETH from `RootBridgeAgent.sol` to `_router`,

After that, we will arrive at [MulticallRootRouter.executeSignedDepositSingle](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/MulticallRootRouter.sol#L312-L389), as [MulticallRootRouter.executeSignedDepositSingle's](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/MulticallRootRouter.sol#L312C5-L389) declaration, `userAccount` will still be Alice's VirtualAccount address. 

Then if we go to [MulticallRootRouter.sol#L331-L351](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/MulticallRootRouter.sol#L331-L351), `_approveAndCallOut` is called at [MulticallRootRouter.sol#L343-L351](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/MulticallRootRouter.sol#L343-L351)

According to [MulticallRootRouter._approveMultipleAndCallOut](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/MulticallRootRouter.sol#L544-L575), the first paramter `refundee` is **address to refund excess gas to**.
And by going through all the code above, the `refundee` will be Alice's Virtual Account.
```solidity
    function _approveMultipleAndCallOut(
        address refundee,
        address recipient,
        address[] memory outputTokens,
        uint256[] memory amountsOut,
        uint256[] memory depositsOut,
        uint16 dstChainId,
        GasParams memory gasParams
    ) internal virtual {
        // Save bridge agent address to memory
        address _bridgeAgentAddress = bridgeAgentAddress;

        // For each output token
        for (uint256 i = 0; i < outputTokens.length;) {
            // Approve Root Port to spend output hTokens.
            outputTokens[i].safeApprove(_bridgeAgentAddress, amountsOut[i]);
            unchecked {
                ++i;
            }
        }

        //Move output hTokens from Root to Branch and call 'clearTokens'.
        IBridgeAgent(_bridgeAgentAddress).callOutAndBridgeMultiple{value: msg.value}(
            payable(refundee),
            recipient,
            dstChainId,
            "",
            SettlementMultipleInput(outputTokens, amountsOut, depositsOut),
            gasParams,
            true
        );
    }
```

So to sum up:
1. If the protocol have ETH within the pool, then Alice's tx will steal all the ETH
1. If the protocol doesn't have any ETH, then Alice's tx will revert

## Tools Used
VIM

## Assessed type

Other
